File: typescript/src/questions/chapter1.ts
/**
 * 1.1 Is Unique: Implement an algorithm to determine if a string has all unique characters.
 * Assume only lower case letters can be used.
 * What if you cannot use additional data structures?
 */

/** Time: O(n), max 26 iterations. Space: O(n), max 26 map entries. */
export function isUnique(str: string) {
  if (str.length > 26) {
    return false;
  }
  const charMap: Record<string, boolean> = {};
  for (let i = 0; i < str.length; i++) {
    const char = str.charAt(i);
    if (charMap[char]) {
      return false;
    }
    charMap[char] = true;
  }
  return true;
}

/** Time: O(n), max 26 iterations. Space: O(1). */
export function isUniqueBits(str: string) {
  if (str.length > 26) {
    return false;
  }
  // Substitute the charMap from the previous solution with a bit vector, represented by an int.
  let bitVector = 0; // 32 bit in JavaScript.
  for (let i = 0; i < str.length; i++) {
    // Subtract by "a" ASCII code to get a value between 0 and 25.
    const n = str.charCodeAt(i) - "a".charCodeAt(0);
    // 1 << n creates a bit vector where the only 1 bit is the nth bit.
    if ((bitVector & (1 << n)) > 0) { // This can only happen if the nth bit is set in the vector.
      return false;
    }
    bitVector = bitVector | (1 << n); // Set the nth bit in the vector.
  }
  return true;
}

/**
 * 1.2 Check Permutation: Given two strings, write a method to decide if one is a permutation of the other.
 * Assume that case and whitespace are both significant.
 */

/** Time: O(n). Space: O(n). O(nlogn) time and O(1) space can be achieved by sorting and comparing. */
export function isPermutation(a: string, b: string) {
  if (a.length !== b.length) {
    return false;
  }
  const charMap: Record<string, number> = {};
  [...a].forEach(letter => {
    if (!charMap[letter]) {
      charMap[letter] = 0;
    }
    charMap[letter]++;
  });
  for (let i = 0; i < b.length; i++) {
    const letter = b.charAt(i);
    if (charMap[letter] === undefined) { // Explicitly check for undefined since 0 is a possible value.
      return false;
    }
    charMap[letter]--;
    if (charMap[letter] < 0) {
      return false;
    }
  }
  return true;
}

/**
 * 1.3 URLify: Replace all spaces in a string with "%20".
 * Assume no leading whitespaces.
 * This implementation is trivial in TypeScript, so assume c-style fixed size character arrays.
 * The fixed size char array has enough space for the final url string.
 * The length of the actual string is given by trueLength.
 */

/** Time: O(n). Space: O(1). */
export function urlify(url: string[], trueLength: number) {
  const spaceCount = url.slice(0, trueLength)
    .reduce((acc, letter) => acc + (letter === " " ? 1 : 0), 0);
  let index = trueLength + spaceCount * 2;
  for (let i = trueLength - 1; i >= 0; i--) {
    if (url[i] === " ") {
      url[index - 3] = "%";
      url[index - 2] = "2";
      url[index - 1] = "0";
      index -= 3;
    } else {
      url[index - 1] = url[i];
      index--;
    }
  }
}

/**
 * 1.4 Palindrome Permutation: Given a string, write a function to check if it is a permutation of a palindrome.
 * Assume that case and whitespace do not matter, and that there will be no leading or trailing whitespaces.
 * Assume that only letters of the english alphabet will be used.
 */

/** Time: O(n). Space: O(n), arguably O(1) since the map cannot contain more than 26 entries. */
export function isPermutationOfPalindrome(str: string) {
  const normalized = str.toLowerCase();
  // A palindrome with an even number of letters must have no letters with odd frequencies.
  // A palindrome with an odd number of letters must have exactly one letter with an odd frequency.
  let countOdd = 0;
  // Contains the frequency of each character in the string.
  const charFrequency: Record<string, number> = {};
  [...normalized].forEach(letter => {
    if (!charFrequency[letter]) {
      charFrequency[letter] = 0;
    }
    charFrequency[letter]++;
    if (charFrequency[letter] % 2 === 1) { // Odd frequency for this letter.
      countOdd++;
    } else {
      countOdd--;
    }
  });
  return countOdd <= 1;
}

/**
 * 1.5 One Edit Away: There are three types of edits that can be performed on strings: insert
 * a character, remove a character, or replace a character. Given two strings, write a function
 * to check if they are one or less edits away from eachother.
 *
 * Examples:
 * pale, ple -> true (insert or remove)
 * pales, pale -> true (insert or remove)
 * pale, bale -> true (replace)
 * pale, pac -> false
 */

/** Time: O(n). Space: O(1). */
export function oneEditAway(a: string, b: string) {
  if (Math.abs(a.length - b.length) > 1) {
    return false;
  }

  // Insert and remove can be seen as one operation.
  // Adding one to the shorter string is equivalent to removing the extra from the longer string.
  const shorter = a.length < b.length ? a : b;
  const longer = a.length < b.length ? b : a;
  let shorterIndex = 0;
  let longerIndex = 0;
  let foundEdit = false; // This flag can only be flipped once. Return false if flipped again.

  while (shorterIndex < shorter.length && longerIndex < longer.length) {
    if (shorter.charAt(shorterIndex) !== longer.charAt(longerIndex)) { // Edit here.
      if (foundEdit) { // Already made an edit.
        return false;
      }
      foundEdit = true;

      if (shorter.length === longer.length) { // Edit is replacing.
        shorterIndex++;
      }
    } else {
      shorterIndex++; // If no edit needed, increment shorter pointer.
    }
    longerIndex++;
  }
  return true;
}

/**
 * 1.6 String Compression: Implement a function to perform basic string compression using
 * the counts of repeated characters. If the compressed string would not be smaller than
 * the original, return the original instead.
 * Assume that only letters of the english alphabet will be used.
 *
 * Examples:
 * aabbbcddd -> a2b3c1d3
 * aabcde -> aabcde (not a2b1c1d1e1)
 */

/**
 * Time: O(n).
 * Space: O(1). The spread operator is syntactic sugar and does not allocate extra space.
 */
export function compressString(str: string) {
  let count = 0;
  let compressed = "";
  [...str].forEach((letter, index) => {
    count++;
    if (index + 1 >= str.length || letter !== str.charAt(index + 1)) {
      compressed += `${letter}${count}`;
      count = 0;
    }
  });
  return compressed.length < str.length ? compressed : str;
}

/**
 * 1.7 Rotate Matrix: Given an NxN matrix, rotate the contents clockwise 90 degrees in place.
 * Assume the matrix contains integers.
 */

/** Time: O(N^2). Space: O(1). */
export function rotateMatrix(matrix: number[][]) {
  const n = matrix.length;
  for (let layer = 0; layer < n / 2; layer++) {
    const first = layer;
    const last = n - 1 - layer;
    for (let startRotate = first; startRotate < last; startRotate++) {
      const offset = startRotate - first;
      const temp = matrix[first][startRotate];

      // Starting from the left side, rotate clockwise.
      matrix[first][startRotate] = matrix[last - offset][first];
      matrix[last - offset][first] = matrix[last][last - offset];
      matrix[last][last - offset] = matrix[startRotate][last];
      matrix[startRotate][last] = temp;
    }
  }
}

/**
 * 1.8 Zero Matrix: Write a function that transforms a given MxN (M, N > 0) matrix in place;
 * if an element is 0, then that element's entire row and column should be set to 0.
 * Assume that the matrix contains integers only.
 */

/** Time: O(MN). Space: O(M + N). */
export function transformMatrixZeros(matrix: number[][]) {
  const rowFlags = matrix.map(_ => false);
  const columnFlags = matrix[0].map(_ => false);

  const setRowZeros = (rowIndex: number) => {
    matrix[0].forEach((_, columnIndex) => matrix[rowIndex][columnIndex] = 0);
  };
  const setColumnZeros = (columnIndex: number) => {
    matrix.forEach((_, rowIndex) => matrix[rowIndex][columnIndex] = 0);
  };

  matrix.forEach((row, rowIndex) => {
    matrix[0].forEach((_, columnIndex) => {
      if (row[columnIndex] === 0) {
        rowFlags[rowIndex] = true;
        columnFlags[columnIndex] = true;
      }
    });
  });
  rowFlags.forEach((flag, rowIndex) => {
    if (flag) {
      setRowZeros(rowIndex);
    }
  });
  columnFlags.forEach((flag, columnIndex) => {
    if (flag) {
      setColumnZeros(columnIndex);
    }
  });
}

/**
 * 1.9 String Rotation: Given a method isSubstring(s1, s2) which checks if s2 is a substring of
 * s1, write a function that checks if s2 is a rotation of s1 using only one call to isSubstring(s1, s2).
 * Assume that the strings do not contain whitespaces.
 *
 * Example:
 * helloworld is a rotation of orldhellow (and vice versa).
 */

/** Time: O(n), assuming isSubstring has a time complexity of O(n). Space: O(1). */
export function isRotation(s1: string, s2: string) {
  // Implementation is trivial.
  const isSubstring = (s1: string, s2: string) => {
    return s1.includes(s2);
  };
  if (s1.length === s2.length && s1.length > 0) {
    return isSubstring(s1 + s1, s2);
  }
  return false;
}

File: go/questions/chapter1.go
package questions

import (
	"math"
	"strconv"
	"strings"
)

/** See chapter1.ts for problem descriptions. */

// IsUnique returns true if the string contains only unique characters.
// Time: O(n), max 26 iterations.
// Space: O(n), max 26 map entries.
func IsUnique(str string) bool {
	if len(str) > 26 {
		return false
	}
	charMap := make(map[rune]bool)
	for _, char := range str {
		if charMap[char] {
			return false
		}
		charMap[char] = true
	}
	return true
}

// IsUniqueBits implements IsUnique using a bit vector.
// Time: O(n), max 26 iterations.
// Space: O(1).
func IsUniqueBits(str string) bool {
	if len(str) > 26 {
		return false
	}
	bitVector := 0
	for _, char := range str {
		val := uint(char) - uint('a')
		if (bitVector & (1 << val)) > 0 {
			return false
		}
		bitVector = bitVector | (1 << val)
	}
	return true
}

// IsPermutation returns true if one string is a permutation of the other.
// Time: O(n). Space: O(n). O(nlogn) time and O(1) space can be achieved by sorting and comparing.
func IsPermutation(a string, b string) bool {
	if len(a) != len(b) {
		return false
	}
	charMap := make(map[rune]int)
	for _, char := range a {
		// No null check since the default value of an int map is 0.
		charMap[char]++
	}
	for _, char := range b {
		charMap[char]--
		if charMap[char] < 0 {
			return false
		}
	}
	return true
}

// URLify replaces all spaces in a string with "%20".
// Time: O(n). Space: O(1).
func URLify(url []rune, trueLength int) {
	actual := url[:trueLength]
	spaceCount := 0
	for _, char := range actual {
		if char == ' ' {
			spaceCount++
		}
	}
	index := trueLength + spaceCount*2
	for i := trueLength - 1; i >= 0; i-- {
		if url[i] == ' ' {
			url[index-3] = '%'
			url[index-2] = '2'
			url[index-1] = '0'
			index -= 3
		} else {
			url[index-1] = url[i]
			index--
		}
	}
}

// IsPermutationOfPalindrome returns true if the string is the permutation of a palindrome.
// Time: O(n). Space: O(n).
func IsPermutationOfPalindrome(str string) bool {
	normalized := strings.ToLower(str)
	// A palindrome with an even number of letters must have no letters with odd frequencies.
	// A palindrome with an odd number of letters must have exactly one letter with an odd frequency.
	countOdd := 0
	// Contains the frequency of each character in the string.
	charFrequency := make(map[rune]int)
	for _, char := range normalized {
		charFrequency[char]++
		if charFrequency[char]%2 == 1 {
			countOdd++
		} else {
			countOdd--
		}
	}
	return countOdd <= 1
}

// IsPermutationOfPalindromeBits implements the above solution using a bit vector.
// Time: O(n). Space: O(1).
func IsPermutationOfPalindromeBits(str string) bool {
	normalized := strings.ToLower(str)
	// Substitute the frequency map from the previous solution with a bit vector.
	bitVector := 0
	// Toggle the nth bit in a bit vector.
	toggleBit := func(vector int, n uint) int {
		if (vector & (1 << n)) == 0 { // The nth bit is not set.
			return vector | (1 << n) // Set the nth bit.
		}
		return vector & ^(1 << n) // Unset the nth bit.
	}
	for _, char := range normalized {
		bitVector = toggleBit(bitVector, uint(char)-uint('a'))
	}
	// There can be at most one set bit if the string is a palindrome permutation.
	return (bitVector == 0) || ((bitVector & (bitVector - 1)) == 0)
}

// OneEditAway returns true if one string is one character edit away from another.
// Time: O(n). Space: O(1).
func OneEditAway(a string, b string) bool {
	if math.Abs(float64(len(a))-float64(len(b))) > 1 {
		return false
	}
	// Insert and remove can be seen as one operation.
	// Adding one to the shorter string is equivalent to removing the extra from the longer string.
	var shorter, longer string
	if len(a) < len(b) {
		shorter = a
		longer = b
	} else {
		shorter = b
		longer = a
	}
	shorterIndex, longerIndex := 0, 0
	foundEdit := false

	for shorterIndex < len(shorter) && longerIndex < len(longer) {
		if shorter[shorterIndex] != longer[longerIndex] {
			if foundEdit { // Already made an edit.
				return false
			}
			foundEdit = true

			if len(shorter) == len(longer) { // Edit is replacing.
				shorterIndex++
			}
		} else {
			shorterIndex++ // If no edit needed, increment shorter pointer.
		}
		longerIndex++
	}
	return true
}

// CompressString compresses a string using counts of repeated characters.
// If the compressed string is not smaller than the original, then the original is returned.
// Time: O(n). Space: O(1).
func CompressString(str string) string {
	count := 0
	compressed := ""

	for i := 0; i < len(str); i++ {
		count++
		if i+1 >= len(str) || str[i] != str[i+1] {
			compressed += string(str[i]) + strconv.Itoa(count)
			count = 0
		}
	}
	if len(compressed) < len(str) {
		return compressed
	}
	return str
}

// RotateMatrix rotates the contents of an NxN matrix clockwise 90 degrees in place.
// Time: O(N^2). Space: O(1).
func RotateMatrix(matrix [][]int) {
	n := len(matrix)
	for layer := 0; layer < n/2; layer++ {
		first := layer
		last := n - 1 - layer
		for startRotate := first; startRotate < last; startRotate++ {
			offset := startRotate - first
			temp := matrix[first][startRotate]

			// Starting from the left side, rotate clockwise.
			matrix[first][startRotate] = matrix[last-offset][first]
			matrix[last-offset][first] = matrix[last][last-offset]
			matrix[last][last-offset] = matrix[startRotate][last]
			matrix[startRotate][last] = temp
		}
	}
}

// TransformMatrixZeros transforms an MxN (M, N > 0) matrix in place;
// if an element is 0, then that element's entire row and column should be set to 0.
// Time: O(MN). Space: O(M + N).
func TransformMatrixZeros(matrix [][]int) {
	rowFlags := make([]bool, len(matrix))
	columnFlags := make([]bool, len(matrix[0]))

	setRowZeros := func(rowIndex int) {
		for columnIndex := range matrix[0] {
			matrix[rowIndex][columnIndex] = 0
		}
	}
	setColumnZeros := func(columnIndex int) {
		for rowIndex := range matrix {
			matrix[rowIndex][columnIndex] = 0
		}
	}

	for rowIndex, row := range matrix {
		for columnIndex := range matrix[0] {
			if row[columnIndex] == 0 {
				rowFlags[rowIndex] = true
				columnFlags[columnIndex] = true
			}
		}
	}
	for rowIndex, flag := range rowFlags {
		if flag {
			setRowZeros(rowIndex)
		}
	}
	for columnIndex, flag := range columnFlags {
		if flag {
			setColumnZeros(columnIndex)
		}
	}
}

// IsRotation returns true if s1 is a rotation of s2 or vice versa.
// Given a function isSubstring, implement this with only one call to isSubstring.
// Time: O(n), assuming isSubstring has a time complexity of O(n). Space: O(1).
func IsRotation(s1 string, s2 string) bool {
	// Implementation is trivial.
	isSubstring := func(s1 string, s2 string) bool {
		return strings.Contains(s1, s2)
	}
	if len(s1) == len(s2) && len(s1) > 0 {
		return isSubstring(s1+s1, s2)
	}
	return false
}
